### PR TITLE
[IMP] l10n_ar_txt_tucuman: new module

### PR DESCRIPTION
**Pull request by pablohmontenegro**
Tarea: 38200

### CODE REVIEW

**Comment by roboadhoc** [comment]
[Pull request status dashboard](https://mergebot.adhoc.ar/ingadhoc/odoo-argentina-ee/pull/461).

File: l10n_ar_txt_tucuman/README.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+   :alt: ADHOC SA
+   :target: https://www.adhoc.com.ar
+
+.. |icon| image:: https://raw.githubusercontent.com/ingadhoc/maintainer-tools/master/resources/adhoc-icon.png
+
+.. image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+   :target: https://www.gnu.org/licenses/agpl
+   :alt: License: AGPL-3
+
+=============================
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/README.rst]: corregir

File: l10n_ar_txt_tucuman/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+    'name': 'Txt Tucuman',
+    'version': "16.0.1.0.0",
+    'category': 'Accounting',
+    'website': 'www.adhoc.com.ar',
+    'license': 'LGPL-3',
+    'images': [
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/__manifest__.py]: quitar

File: l10n_ar_txt_tucuman/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    ],
+    'depends': [
+        'l10n_ar_account_tax_settlement',
+    ],
+    'data': [
+    ],
+    'demo': [
+    ],
+    'test': [
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/__manifest__.py]: quitar

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    def iibb_tucuman_files_values(self, move_lines):
+        """ Implementado segun especificación indicada en tarea 38200. Ver especificación también en l10n_ar_txt_tucuman/doc/MRETPER6R2.pdf a partir de la página 12. """
+        self.ensure_one()
+        content_datos = ''
+        content_retper = ''
+        content_ncfact = ''
+
+        if nc_without_reversed_entry_id := move_lines.filtered(lambda x: x.move_type == 'out_refund' and not x.move_id.reversed_entry_id):
+            raise ValidationError(f"Algunos comprobantes rectificativos no contienen información de que comprobante original están revirtiendo: %s" % (", ".join(nc_without_reversed_entry_id.mapped('move_id.name'))))
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/models/account_journal.py]:
```suggestion
            raise UserError(_("Algunos comprobantes rectificativos no contienen información de que comprobante original están revirtiendo: %s") % (", ".join(nc_without_reversed_entry_id.mapped('move_id.name'))))
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        if moves_without_street_city_state := move_lines.filtered(lambda x: not x.partner_id.street or not x.partner_id.city or not x.partner_id.state_id or not x.partner_id.zip):
+            raise ValidationError(f"Algunos comprobantes no contienen información acerca de la calle/ciudad/provincia/cod postal del contacto: %s" % (", ".join(moves_without_street_city_state.mapped('move_id.name'))))
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/models/account_journal.py]:
```suggestion
            raise UserError(_("Algunos comprobantes no contienen información acerca de la calle/ciudad/provincia/cod postal del contacto: %s") % (", ".join(moves_without_street_city_state.mapped('move_id.name'))))
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
+from odoo import models, fields
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/models/account_journal.py]:
```suggestion
from odoo import models, fields, _
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+from odoo.exceptions import ValidationError
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/models/account_journal.py]:
```suggestion
from odoo.exceptions import UserError
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        for line in move_lines.sorted(key=lambda r: (r.date, r.id)):
+            # Archivo DATOS.TXT
+            # FECHA, longitud: 8. Formato AAAAMMDD
+            content_datos += fields.Date.from_string(line.date).strftime('%Y%m%d')
+            # TIPODOC, longitud: 2
+            content_datos += line.partner_id.l10n_latam_identification_type_id.l10n_ar_afip_code
+            # DOCUMENTO, longitud: 11
+            content_datos += line.partner_id.l10n_ar_vat
+            # TIPO COMP, longitud: 2
+            # 99 para retenciones por el ejemplo que pasó en el archivo adjunto el cliente en la tareas
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/models/account_journal.py]:
```suggestion
            # 99 para retenciones por el ejemplo que pasó en el archivo adjunto el cliente en la tarea
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        ('iibb_tucuman', 'TXT Retenciones/Percepciones Tucuman')
+    ])
+
+    def iibb_tucuman_files_values(self, move_lines):
+        """ Implementado segun especificación indicada en tarea 38200. Ver especificación también en l10n_ar_txt_tucuman/doc/MRETPER6R2.pdf a partir de la página 12. """
+        self.ensure_one()
+        content_datos = ''
+        content_retper = ''
+        content_ncfact = ''
+
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/models/account_journal.py:16]: Agragar validación para que no tenga punto de venta de 5 dígitos

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            content_datos += line.move_id.l10n_latam_document_type_id.l10n_ar_letter if line.l10n_latam_document_type_id.internal_type == 'invoice' else ' '
+            # COD. LUGAR EMISION, longitud: 4
+            pos, number = get_pos_and_number(line.move_id.l10n_latam_document_number)
+            content_datos += pos[-4:]
+            # NUMERO, longitud: 8
+            content_datos += number
+            # BASE_CALCULO, longitud: 15,2
+            content_datos += '%015.2f' % (line.tax_base_amount if line.l10n_latam_document_type_id.internal_type == 'invoice' else line.payment_id.withholding_base_amount)
+            # PORCENTAJE/ALICUOTA, longitud: 6,3
+            content_datos += '%06.3f' % line.tax_line_id.get_partner_alicuot(line.partner_id, line.date).alicuota_percepcion
```
Comment (pablohmontenegro) [inline l10n_ar_txt_tucuman/models/account_journal.py]: que tome alicuota_retencion para retenciones

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        content_retper += ' ' * 11
+        # 9, C. POSTAL, longitud: 8
+        content_retper += '%8s' % line.partner_id.zip
+        content_retper += '\r\n'
+        return content_retper if content_retper else ''
+
+    def _iibb_tucuman_ncfact_txt_file(self, line):
+        # Archivo NCFACT.TXT
+        content_ncfact = ''
+        if line.move_type == 'out_refund':
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py]: 1. agregar doscstring.
2. recibir lineas (todas las que son necesarias para crear el archivo)

```python

    def _iibb_tucuman_ncfact_txt_file(self, lines):
       """ Devuelve contenido del archivo NCFACT.TXT Txt Tucuman
       """
        content_ncfact = ''
        for line in lines.filtered(lamba x: x.move_type == 'out_refund'):
             ...

```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+            line.move_id.reversed_entry_id.l10n_latam_document_number,
+            line.move_id.reversed_entry_id.l10n_latam_document_type_id.code)
+            content_ncfact += str(document_number_parts['point_of_sale']).zfill(4)
+            # 4, NUMERO FAC, longitud: 8
+            content_ncfact += str(document_number_parts['invoice_number']).zfill(8)
+            # 5, TIPO FAC, longitud: 2
+            content_ncfact += line.move_id.reversed_entry_id.l10n_latam_document_type_id.code.zfill(2)
+            content_ncfact += '\r\n'
+            return content_ncfact
+        return ''
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py]: always return at the end of the method
```suggestion
    return content_ncfact
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+            # 3, COD LUGAR EMISION FAC, longitud: 4
+            document_number_parts = line.move_id._l10n_ar_get_document_number_parts(
+            line.move_id.reversed_entry_id.l10n_latam_document_number,
+            line.move_id.reversed_entry_id.l10n_latam_document_type_id.code)
+            content_ncfact += str(document_number_parts['point_of_sale']).zfill(4)
+            # 4, NUMERO FAC, longitud: 8
+            content_ncfact += str(document_number_parts['invoice_number']).zfill(8)
+            # 5, TIPO FAC, longitud: 2
+            content_ncfact += line.move_id.reversed_entry_id.l10n_latam_document_type_id.code.zfill(2)
+            content_ncfact += '\r\n'
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py:143]: this should end be per line, or per file?

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        if line in percepciones:
+            content_datos += '%06.3f' % partner_alicuot.alicuota_percepcion
+        else:
+            content_datos += '%06.3f' % partner_alicuot.alicuota_retencion
+        # 10, MONTO_RET/PER, longitud: 15,2
+        content_datos += '%015.2f' % abs(line.balance)
+        content_datos += '\r\n'
+        return content_datos if content_datos else ''
+
+    def _iibb_tucuman_retper_txt_file(self, line):
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py]: process lines, return only content of the file. str

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+        # 9, PORCENTAJE/ALICUOTA, longitud: 6,3
+        partner_alicuot = line.tax_line_id.get_partner_alicuot(line.partner_id, line.date)
+        if line in percepciones:
+            content_datos += '%06.3f' % partner_alicuot.alicuota_percepcion
+        else:
+            content_datos += '%06.3f' % partner_alicuot.alicuota_retencion
+        # 10, MONTO_RET/PER, longitud: 15,2
+        content_datos += '%015.2f' % abs(line.balance)
+        content_datos += '\r\n'
+        return content_datos if content_datos else ''
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py]: you already initialize the content_ncfact = ''
```suggestion
        return content_datos
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            raise UserError(_("Algunos comprobantes tienen punto de venta de 5 dígitos y deben tener de 4 dígitos para "
+                              "poder generar el archivo txt de retenciones y percepciones de Tucuman: %s") %
+                            (", ".join(move_lines_with_five_digits_pos.mapped('move_id.name'))))
+        percepciones = move_lines.filtered(lambda x: x.l10n_latam_document_type_id.internal_type == 'invoice')
+        if percepciones and len(percepciones) != len(move_lines):
+            raise UserError(_("Debe generar archivos para TXT Tucuman por separado para retenciones por un lado y "
+                              "percepciones por otro."))
+        return percepciones if percepciones else self.env['account.move.line']
+
+    def _iibb_tucuman_datos_txt_file(self, line, percepciones):
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py]: receive lines instad of lines, do not pass perceptions.
in the iteration of each line create a temporal varialbe is_perception

```python
for line in lines:
    is_perception = line.l10n_latam_document_type_id.internal_type == 'invoice'
    ....
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+            content_ncfact += self._iibb_tucuman_ncfact_txt_file(line)
+        return [{
+            'txt_filename': 'DATOS.txt',
+            'txt_content': content_datos,
+            },
+            {'txt_filename': 'RETPER.txt',
+            'txt_content': content_retper,
+            },
+            {'txt_filename': 'NCFACT.TXT',
+            'txt_content': content_ncfact,
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py]:
```suggestion
        # VALIDACIONES
        self._iibb_tucuman_validations(move_lines)

        # ELABORACIÓN DE ARCHIVOS TXT
        lines = move_lines.sorted(key=lambda r: (r.date, r.id))
        return [{
                'txt_filename': 'DATOS.txt',
                'txt_content': self._iibb_tucuman_datos_txt_file(lines),
            }, {
                'txt_filename': 'RETPER.txt',
                'txt_content': self._iibb_tucuman_retper_txt_file(lines),
            }, {
                'txt_filename': 'NCFACT.TXT',
                'txt_content': self._iibb_tucuman_ncfact_txt_file(lines),
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+            }]
+
+    def _iibb_tucuman_validations(self, move_lines):
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py:33]: Validation should be only validation, not need to return perceptions

**Review by zaoral — CHANGES_REQUESTED** [review]

File: l10n_ar_txt_tucuman/models/account_journal.py (hunk 13 above)
Comment (pablohmontenegro) [reply l10n_ar_txt_tucuman/models/account_journal.py:143]: yes

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    ])
+
+    def iibb_tucuman_files_values(self, move_lines):
+        """ Implementado segun especificación indicada en tarea 38200.
+        Ver especificación también en l10n_ar_txt_tucuman/doc/MRETPER6R2.pdf a partir de la página 12. """
+        self.ensure_one()
+        content_datos, content_retper, content_ncfact = '', '', ''
+
+        # VALIDACIONES
+        percepciones = self._iibb_tucuman_validations(move_lines)
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py]:
```suggestion
        self._iibb_tucuman_validations(move_lines)
```

File: l10n_ar_txt_tucuman/models/account_journal.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+
+    settlement_tax = fields.Selection(selection_add=[
+        ('iibb_tucuman', 'TXT Retenciones/Percepciones Tucuman')
+    ])
+
+    def iibb_tucuman_files_values(self, move_lines):
+        """ Implementado segun especificación indicada en tarea 38200.
+        Ver especificación también en l10n_ar_txt_tucuman/doc/MRETPER6R2.pdf a partir de la página 12. """
+        self.ensure_one()
+        content_datos, content_retper, content_ncfact = '', '', ''
```
Comment (zaoral) [inline l10n_ar_txt_tucuman/models/account_journal.py]:
```suggestion
```

**Review by zaoral — CHANGES_REQUESTED** [review]

**Review by zaoral — APPROVED** [review]

**Comment by zaoral** [comment]
@roboadhoc r+

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Comment by fw-bot-adhoc** [comment]
@pablohmontenegro @zaoral this pull request has forward-port PRs awaiting action (not merged or closed):
- ingadhoc/odoo-argentina-ee#484

**Pull request closed without merging**